### PR TITLE
Assignment3

### PR DESCRIPTION
**Pull request by Eismcsquared**

### CODE REVIEW

File: src/container/ParticleContainer.h [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+     * The getter for the particles in a particle container that are still in domain.
+     * @return The particles in the container as a vector.
+     */
+    std::vector<Particle>& getParticles() const;
+
+    /**
+     *
+     * @param f
+     */
+    void setF(std::unique_ptr<Force> &f);
```
Comment (manishmishra6016) [inline src/container/ParticleContainer.h:79]: This is nothing major, I just see force as something decoupled from container (which it is in someway done here but more in the code itself), and they would more nicely come together while creating the simulation class. I would say you can leave it as it for now but just wanted to leave a comment anyway.

File: src/container/ParticleContainer.h [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+/**
+ * @brief The abstract class that represents a container for particles. Subclasses implement the concrete way to manage particles.
+ */
+class ParticleContainer {
+
+protected:
+/**
+ * Particles stored in the container
+ */
+    std::unique_ptr<std::vector<Particle>> particles;
```
Comment (manishmishra6016) [inline src/container/ParticleContainer.h:54]: Could you maybe comment on why you made this a pointer instead of just using vector<particles>? It seems like this is the main container where you store particles, so it feels like storing them directly would just work.

File: src/container/LinkedCellContainer.cpp [hunk 3]
```diff
@@ -0,0 +1,390 @@
...
+                (*particles)[pointCellParticles[k]].setF(v2_force);
+            }
+        }
+
+        
+    }
+
+}
+
+void LinkedCellContainer::updateCellF(const std::vector<int> &v1, const std::vector<int> &v2, bool newton3){
```
Comment (manishmishra6016) [inline src/container/LinkedCellContainer.cpp:116]: I find the name `updateCellF()` a bit confusing. I understand that it comes from updateF, so maybe updateFCells?
But this is just me, feel free to ignore :D

File: src/container/LinkedCellContainer.cpp [hunk 4]
```diff
@@ -0,0 +1,390 @@
...
+                std::array<double, 3> v1_force;
+                std::array<double, 3> v2_force;
+
+                for(int k = 0; k < 3; k++){
+                    v1_force[k] = (*particles)[i].getF()[k] + forceJI[k];
+                    v2_force[k] = (*particles)[j].getF()[k] + forceIJ[k];
+                }
+
+                (*particles)[i].setF(v1_force);
+                (*particles)[j].setF(v2_force);
```
Comment (manishmishra6016) [inline src/container/LinkedCellContainer.cpp:151]: I don't see any difference in if and else parts. It seems like they are both with newton3. It's fine if you just want to implement this with newton3, but then remove the redundant code.

File: src/container/LinkedCellContainer.cpp [hunk 5]
```diff
@@ -0,0 +1,390 @@
...
+            for (int x = std::max(-1, index3D[0] - 1); x <= std::min(index3D[0] + 1, nCells[0]); ++x) {
+                int neighborIndex = get1DIndex({x, y, z});
+                if (neighborIndex != cellIndex) {
+                    neighbors.push_back(neighborIndex);
+                }
+            }
+        }
+    }
+    return neighbors;
+}
```
Comment (manishmishra6016) [inline src/container/LinkedCellContainer.cpp:196]: so the neighbors of the cells do not change even when particles inside them change, so you can compute these neighbors only once in the beginning and store them in cells instead of computing them each time.

File: src/container/LinkedCellContainer.cpp [hunk 6]
```diff
@@ -0,0 +1,390 @@
...
+    if (!(domainSize == other.domainSize) || cutoff != other.cutoff || !(boundaryConditions == other.boundaryConditions)) {
+        return false;
+    }
+    return true;
+}
+
+void LinkedCellContainer::addParticle(const Particle& particle) {
+    int cellIndex = getCellIndex(particle.getX());
+    if (cellIndex >= 0 && cellIndex < cells.size()) {
+        cells[cellIndex].addIndex((int) particles->size());
```
Comment (manishmishra6016) [inline src/container/LinkedCellContainer.cpp:258]: old C-style casting to int. Instead use ```std::static_cast```

File: src/container/LinkedCellContainer.cpp [hunk 7]
```diff
@@ -0,0 +1,390 @@
...
+    std::array<int, 3> index3D = get3DIndex(index);
+    return index3D[0] < 0 || index3D[0] >= nCells[0] || index3D[1] < 0 || index3D[1] >= nCells[1] || index3D[2] < 0 || index3D[2] >= nCells[2];
+}
+
+std::vector<int> LinkedCellContainer::getAllHaloIndices(Direction direction) {
+    std::vector<int> haloIndices;
+    switch (direction) {
+        case LEFT:
+            for (int z = -1; z <= nCells[2] ; ++z) {
+                for (int y = -1; y <= nCells[1]; ++y) {
```
Comment (manishmishra6016) [inline src/container/LinkedCellContainer.cpp:286]: these can also be computed once and stored

File: src/container/LinkedCellContainer.cpp [hunk 8]
```diff
@@ -0,0 +1,390 @@
...
+                    (*particles)[p].removeFromDomain();
+                    cells[i].removeIndex(p);
+                    break;
+                case REFLECTING:
+                    // change sign of the corresponding velocity component if the particle is flying away.
+                    if ((*particles)[p].getV()[direction / 2] * (direction % 2 - 0.5) > 0) {
+                        std::array<double, 3> vel = (*particles)[p].getV();
+                        vel[direction / 2] *= -1;
+                        (*particles)[p].setV(vel);
+                    }
```
Comment (manishmishra6016) [inline src/container/LinkedCellContainer.cpp:352]: It seems like the particle stays in the halo cell even though you reverse the velocity. Because you don't shift it back inside the domain, it is not considered during the force computation? Could you clarify this?

File: src/container/LinkedCellContainer.cpp [hunk 9]
```diff
@@ -0,0 +1,390 @@
...
+                    break;
+                case PERIODIC:
+                    cells[i].removeIndex(p);
+                    std::array<double, 3> pos = (*particles)[p].getX();
+                    pos[direction / 2] = pos[direction / 2] + pow(-1, direction % 2) * domainSize[direction / 2];
+                    (*particles)[p].setX(pos);
```
Comment (manishmishra6016) [inline src/container/LinkedCellContainer.cpp:358]: not added to another cell after moving, just pointing out even though this was not necessary for WS3. Also, maybe add comments about the logic.

File: src/body/Cuboid.h [hunk 10]
```diff
@@ -0,0 +1,26 @@
...
+ */
+class Cuboid : public Cluster {
+private:
+    /**
+     * The number of particles in x, y and z direction.
+     */
+    std::array<unsigned int, 3> N;
+
+public:
+    Cuboid(const std::array<double, 3> &x, const std::array<double, 3> &v, const std::array<unsigned int, 3> &n,
```
Comment (manishmishra6016) [inline src/body/Cuboid.h:16]: maybe add a comment about `x` that it is bottom-left corner or something like that? because this `x` in cluster will store different information for cuboid and sphere, right?

File: tests/SphereTest.cpp [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+    EXPECT_NEAR(2, meanSquire, 5e-3) << "Wrong mean squared velocity of Brownian motion. Expected: "
+    << 2 << ", but got " << meanSquire;
+    if(::testing::Test::HasFailure()) {
+        test_logger->info("Sphere - 2D Sphere test failed");
+    } else {
+        test_logger->info("Sphere - 2D Sphere test passed");
+    }
+}
+
+// Test whether a 2D sphere generates the correct particles in terms their position, mass and velocity including Brownian motion.
```
Comment (manishmishra6016) [inline tests/SphereTest.cpp:42]:
```suggestion
// Test whether a 3D sphere generates the correct particles in terms their position, mass and velocity including Brownian motion.
```

File: tests/SphereTest.cpp [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+    const double pi = 3.141592653589793;
+};
+
+// Test whether a 2D sphere generates the correct particles in terms their position, mass and velocity including Brownian motion.
+TEST_F(SphereTest, CreateParticles2D) {
+    test_logger->info("Sphere - 2D Sphere test");
+    std::vector<Particle> particles;
+    sphere2D.createParticles(particles);
+    double mean = 0;
+    double meanSquire = 0;
```
Comment (manishmishra6016) [inline tests/SphereTest.cpp:21]:
```suggestion
    double meanSquare = 0;
```

File: tests/SphereTest.cpp [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+    for(Particle &p: particles) {
+        EXPECT_TRUE(ArrayUtils::L2Norm(p.getX()) <= 1000) << "Wrong particle created at position ("
+        << p.getX()[0] << ", " << p.getX()[1] << ", " << p.getX()[2] << ")";
+        EXPECT_NEAR(1, p.getM(), 1e-12) << "Particle with wrong mass encountered. Expected: 1, but got: " << p.getM();
+        mean += ArrayUtils::L2Norm(p.getV());
+        meanSquire += pow(ArrayUtils::L2Norm(p.getV()), 2);
```
Comment (manishmishra6016) [inline tests/SphereTest.cpp:27]:
```suggestion
        meanSquare += pow(ArrayUtils::L2Norm(p.getV()), 2);
```

File: tests/SphereTest.cpp [hunk 14]
```diff
@@ -0,0 +1,67 @@
...
+    EXPECT_NEAR(sqrt(2 / pi), mean, 1e-3) << "Wrong mean velocity of Brownian motion. Expected: "
+    << sqrt(2 / pi) << ", but got " << mean;
+    EXPECT_NEAR(0.75, meanSquire, 2.5e-3) << "Wrong mean squared velocity of Brownian motion. Expected: "
+    << 0.75 << ", but got " << meanSquire;
+    if(::testing::Test::HasFailure()) {
+        test_logger->info("Sphere - 3D Sphere test failed");
+    } else {
+        test_logger->info("Sphere - 3D Sphere test passed");
+    }
+}
```
Comment (manishmishra6016) [inline tests/SphereTest.cpp:67]: one natural test that comes to mind is to see if all the particles created are actually within the radius, what do you think?

File: tests/LinkedCellContainerTest.cpp [hunk 15]
```diff
@@ -0,0 +1,220 @@
...
+    if (::testing::Test::HasFailure()) {
+        test_logger->info("LinkedCellContainer - Reflecting test failed");
+    } else {
+        test_logger->info("LinkedCellContainer - Reflecting test passed");
+    }
+}
+
+// Test whether the potential is correctly cut of at the cutoff radius on a two body system. The idea is, after both particles
+// leaving the potential of each other, their kinetic energy should be the sum of their initial kinetic energy plus the potential
+// they pass throw.
```
Comment (manishmishra6016) [inline tests/LinkedCellContainerTest.cpp:200]: nice idea!

**Review by manishmishra6016 — APPROVED** [review]
**Note**
_This is only feedback regarding the code directly. Grade feedback will be on Moodle once we are through with all groups._

-----------------------------------------------------------------------------------------------------

Thank you for your great submission 💪
I have no major complains, you guys did a great job. Have a look at my comments, respond to the ones I have requested clarification on soon, and the rest you can see if they make sense to you and you can implement some changes. But ofcourse, first focus on the worksheet 4 tasks. I am sure you will do great.

Good luck! 😄

File: src/container/LinkedCellContainer.cpp (hunk 8 above)
Comment (Eismcsquared) [reply src/container/LinkedCellContainer.cpp:352]: We actually don't distinguish halo cells from normal cells in the force calculation, so particles in the halo cells are also considered (In case of outflow, particles are immediately removed from halo cells). By flipping the sign, the particle will most likely return to domain in the next time step. Ofcourse, there is potentially a problem if the force is too large as the update of x contains a force term. However, since this term scales as delta_t^2, choosing delta_t sufficiently small should avoid this problem. Nevertheless, I will consider making it more stable.

File: src/container/LinkedCellContainer.cpp (hunk 4 above)
Comment (Eismcsquared) [reply src/container/LinkedCellContainer.cpp:151]: There is a difference. In case of newton3==false, the force is computed twice. I actually would like to remove this part as well, but I thought we must keep the possibility to do simulations without Newton 3. Good to know that we don't, I will completely remove the newton3 flag.

File: tests/SphereTest.cpp (hunk 12 above)
Comment (Eismcsquared) [reply tests/SphereTest.cpp:21]: I don't see what is the suggested change here...

File: tests/SphereTest.cpp (hunk 13 above)
Comment (Eismcsquared) [reply tests/SphereTest.cpp:27]: And here also

File: tests/SphereTest.cpp (hunk 14 above)
Comment (Eismcsquared) [reply tests/SphereTest.cpp:67]: This is indeed tested in line 50

File: tests/SphereTest.cpp (hunk 11 above)
Comment (Eismcsquared) [reply tests/SphereTest.cpp:42]: oops, forgot to change after copying

File: src/container/ParticleContainer.h (hunk 2 above)
Comment (Eismcsquared) [reply src/container/ParticleContainer.h:54]: The thing is I don't want to create a empty vector in the main function and pass it to the XML reader, because adding an extra parameter just for passing an empty vector does not seem to be a good idea for me (or is this actually better?). I want to create the vector in the function for reading xml instead. To make sure that the memory for storing particles remains valide during the entire program, there only seem to be two options: either storing the particles directly (not as reference) in the container, which means we have to copy all the particles into it, or storing them on the heap. To do the latter, the only possibility seems to be creating pointers. Since smart pointers automatically free the memory once they go out of scope, I have to store them as attributes (right?), because if I just store a reference to the vector on the heap, the vector will automatically be destructed.

File: src/container/LinkedCellContainer.cpp (hunk 9 above)
Comment (Eismcsquared) [reply src/container/LinkedCellContainer.cpp:358]: Can we generally assume that the boundary conditions in the same direction are always the same (e.g. the left and the right boundary always have the same boundary condition)? because it does not really make sense to have a periodic boundary on the left hand side and something else on the right hand side.

File: tests/SphereTest.cpp (hunk 12 above)
Comment (manishmishra6016) [reply tests/SphereTest.cpp:21]: just minor spelling :D

File: src/container/LinkedCellContainer.cpp (hunk 9 above)
Comment (manishmishra6016) [reply src/container/LinkedCellContainer.cpp:358]: It is only always true for periodic boundary, not for others. For others, you have to consider separately for each boundary. So you can indeed have cases where left is reflective while right is outflow.

File: tests/SphereTest.cpp (hunk 12 above)
Comment (Eismcsquared) [reply tests/SphereTest.cpp:21]: oh yes

File: src/container/ParticleContainer.h (hunk 2 above)
Comment (Eismcsquared) [reply src/container/ParticleContainer.h:54]: Hi, I saw your comment "unnecessary pointers" in the grading, is it bad to store particles as pointers? Do you have any suggestion on this? Also, you mentioned raw pointers, which I think we didn't use. Could you please point out where it is?

Comment (manishmishra6016) [reply src/container/ParticleContainer.h:54]: Sorry I didn't reply to your comment. I think passing an empty vector is practically free and feels like a much nicer way to implement this. I do find use of pointers here a bit unnecessary. About raw pointers, I don't remember if I noticed it somewhere or I made a mistake, in any case, it is nice if you didn't use it :)
